### ISSUE TITLE
Communicate routingMode in events

### ISSUE THREAD

**Issue by sebhoerl** [matsim-org/matsim-libs#1577]
I was just experimenting with some feeder simulations by chaining the DRT routing module and PT routing module. The reuslting "routing mode" is called "feeder". What I noticed is that we can nicely deduct the mode of each leg from the events (through the departure events) but there is not really a way to deduct the "trip mode", which would be the "routingMode" internally of MATSim. So effectively, "feeder" never appears anywhere in the events.

Some options:
- Introduce a new event (e.g. `PersonTripEvent`) that is sent by the QSim when an agent departs from a non-interaction activity
- Add a `routingMode` attribute to the `PersonDepartureEvent`

What do you think? (I'd find the second option less intrusive)

**Comment by vsp-gleich** [comment]
Since I was and am still working on intermodal routing and routing modes: I think that depends also on where we intend to save that information on routing modes. That opens up to me also the question what your use case is (besides that it might be nice to know anyway).
There are some old discussions on atlassian about implementing a trip element (https://matsim.atlassian.net/browse/MATSIM-847) which unfortunately does not seem to be the preferred solution for all and might never come although it seems cleaner to me (save the routing mode only once per trip). So to me it seems somewhat awkward to have an PersonTripEvent when we have no trip element we could associate with it. 
Add a routingMode attribute to the PersonDepartureEvent seems more straightforward to me as long as we remain with saving the routing mode in the legs.

**Comment by sebhoerl** [comment]
Yes, my concern is only for the events. Currently, having only an events file, it is not really possible to figure out what routingMode a certain trip had (even after reconstructing the legs belonging to a specific trip).

Based on the comments and likes, I guess going with an attribute to the PersonDepartureEvent sounds like a good way forward.

For the use case: We're currently looking at a feeder service to a rail station. So what I did is to set up an ad-hoc routing module which checks whether a trip starts / ends in the operating area around the station and ends / starts outside of it. In the former case, an access trip will be generated using the DRT RoutingModule and the PT Routing module like `walk -> drt -> walk -> pt -> ...` and in the second case an egress trip is generated like `... -> pt -> walk -> drt -> walk`. Otherwise it gives null (i.e. fallback).

It would be nice to have a more streamlined code base where this can be configured flexibly. Using the access/egress feature of the SRR always seems a bit poor in features for me (1) because we want to run it with discrete choice models, so we explicitly want to generate trips for [mode A + pt] and [mode B + pt] as two independent altenratives, and (2) best-response or random may not be the most apt approaches even with scoring / subtour mode choice, because sometimes we may want to have a [X + pt] mode as a chain-based mode if we think of P+R and similar things. What I would have in mind is some functionality that can create an intermodal "mode" via code or via configuration with additional tools such as an interface for selecting the right interaction place, etc. But these are some future wishes. For the time being, I have the ad-hoc routing module as described above ;)

**Comment by kainagel** [comment]
One problem is that users regularly use the "routingMode" (which is a software implementation thing) instead of the "mainMode" (which is the transportation planning thing).  It would be good if we could keep "routingMode" in a way that we will not have, at some point, to support backwards compatibility because it was used instead of the transportation planning thing.

**Comment by sebhoerl** [comment]
I added a PR, open for discussion. I guess the routingMode vs. mainMode topic refers still to the old "direct transit walk" topic? Is there any other example? From this perspective it may be wise not to make the change, but on the other hand it would simplify analyses at a lot, especially if we look more and more into intermodal use cases. In DMC we usually have a constraint for this which filters out the PT alternative, but with scoring I do see the problem.

**Comment by vsp-gleich** [comment]
I think the routingMode vs. mainMode topic is about people missinterpreting the routing mode as a main mode for a lack of understanding of the difference and lack of care to think about it. I might be in part a culprit, because I left the RoutingModeIdentifier as the default AnalysisMainModeIdentifier because the RoutingModeMainModeIdentifier is the only MainModeIdentifier which will never fail whereas any proper AnalysisMainModeIdentifier either risks missinterpreting intermodal trips (like the old MainModeIdentifierImpl) or fails when stumbling over a new mode not known to it (TranportPlanningMainModeIdentifier would be a good AnalysisMainModeIdentifier but fails for any leg mode not implemented in it, so adding a new mode ends in matsim crashing unless a proper AnalysisMainModeIdentifier is bound). Besides transit walk where routing mode and main mode differ it concerns especially intermodal trips where routing mode tells which modes could have been used and main mode should tell which modes have really been used. I can bind AnalysisMainModeIdentifier by default to TranportPlanningMainModeIdentifier in a separate pull request, but that will not stop people from interpreting the routing mode attribute of that enhanced PersonDepartureEvent as some kind of transport planning main mode. Still having that information helps to understand e.g. how many trips routed with pt or intermodally did actually use pt or intermodal access/egress, so that seems to me a legitimate use case (the alternative is analyzing the plans and somehow mapping to the trips.csv, that's not nice).

**Comment by kainagel** [comment]
Maybe use a different name?  Something like "computationalRoutingMode"?

**Comment by sebhoerl** [comment]
I updated the PR to `computationalRoutingMode` as proposed by Kai. This may be a good trade-off.

**Comment by sebhoerl** [comment]
@vsp-gleich Do you think it would make sense to add `computationalRoutingMode` also to the trips and legs output?

**Comment by vsp-gleich** [comment]
Yes, I can add that to the trips.csv

**Comment by sebhoerl** [comment]
Side note: This will also make it possible to pass on the `routingMode` on to the experienced plans.
